### PR TITLE
expression: support enum pushdown

### PR DESCRIPTION
**Pull request by dyzsr**
### What problem does this PR solve?

Problem Summary:

- This PR enables enum pushdown together with https://github.com/tikv/tikv/pull/9648

### What is changed and how it works?

- Set `EnumSetAsInt` flag for `Enum` types when casting enum as int
- Support Encoding/Decoding of `Enum` type

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

### Release note 

- Support enum pushdown.

### CODE REVIEW

**Comment by dyzsr** [comment]
Waiting for https://github.com/pingcap/parser/pull/1165

**Review by wshwsh12 — COMMENTED** [review]
Please avoid using enum index and add a simple test case.
```
drop table t;
create table t(e enum('c','b','a'));
insert into t values(1),(2),(3);
select e from t where e > 'b';
select e from t where e > 2;

alter table t add index idx(e);
select e from t where e > 'b';
select e from t where e > 2;

```

**Comment by wshwsh12** [comment]
https://github.com/pingcap/tidb/blob/8e073163bca275d181142c2c048c611e4bcc32fb/expression/expression.go#L1156-L1161 
We should add a check to avoid push down Enum expression to TiFlash.

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1844,6 +1844,14 @@ func BuildCastFunction(ctx sessionctx.Context, expr Expression, tp *types.FieldT
 // WrapWithCastAsInt wraps `expr` with `cast` if the return type of expr is not
 // type int, otherwise, returns `expr` directly.
 func WrapWithCastAsInt(ctx sessionctx.Context, expr Expression) Expression {
+	if expr.GetType().Tp == mysql.TypeEnum {
+		if col, ok := expr.(*Column); ok {
+			col = col.Clone().(*Column)
```
Comment (lzmhhh123) [inline expression/builtin_cast.go:1849]: Why also clone col here?

**Review by lzmhhh123 — COMMENTED** [review]
Rest LGTM

File: expression/builtin_cast.go (hunk 1 above)
Comment (wshwsh12) [reply expression/builtin_cast.go:1849]: In the expression `if(e>1, e,'a')`,  the first `e` evalType is Int, the second `e` evalType is String. If don't clone col, the two pointers will point to the same address..

**Review by lzmhhh123 — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]
LGTM

**Pull request merged**